### PR TITLE
Refactor SPSA

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Refactor the SPSA optimizer to allow more flexible options and prepare the structure for extensions such as 2-SPSA and QN-SPSA.

### Details and comments

Most existing code--and all code in the tests--still works exactly as before. 

Changes to before:
* Instead of assuming powerlaws for the perturbation and learning rate, arbitrary sequences are now supported (e.g. constant)
* The coefficients for the powerlaws cannot be passed in anymore, instead a generic iterator must be passed (a convience function is provided for that)

New options:
* A blocking condition: only accept a new step if it improves the loss (modulo some allowed increase)
* Using multiple gradient samples in each step instead of just one
* Perturbing only a few dimensions instead of all
* Setting a final tolerance and if the update step is smaller than the tolerance, stop
* A callback function
* Setting a trust region (update step is confined to the learning rate)

The particular structure of the code allows a straightforward extension to [2-SPSA](https://ieeexplore.ieee.org/document/657661/;jsessionid=Md2DInUJuwng6ogonPq6CCdXa3orze7z8ioeg42edir0S4_P3Kq2!1659866155) and [QN-SPSA](https://arxiv.org/abs/2103.09232), not included in this PR to keep it small.

### CODE REVIEW

File: qiskit/algorithms/optimizers/spsa.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # This code is part of Qiskit.
 #
-# (C) Copyright IBM 2018, 2021.
```
Comment (nonhermitian) [inline qiskit/algorithms/optimizers/spsa.py:3]: why change this?

File: qiskit/algorithms/optimizers/spsa.py [hunk 2]
```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
...
-                      initial_theta: np.ndarray,
-                      maxiter: int,
-                      save_steps: int = 1,
-                      last_avg: int = 1) -> List:
-        """Minimizes obj_fun(theta) with a simultaneous perturbation stochastic
-        approximation algorithm.
+def bernoulli_perturbation(dim, perturbation_dims=None):
+    """Get a Bernoulli random perturbation."""
+    if perturbation_dims is None:
+        return np.array([1 - 2 * np.random.binomial(1, 0.5) for _ in range(dim)])
```
Comment (nonhermitian) [inline qiskit/algorithms/optimizers/spsa.py:363]: why does this not use the `size` kwarg in the `binomial` call?

File: qiskit/algorithms/optimizers/spsa.py [hunk 3]
```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
...
+        # compute the perturbations
+        if isinstance(self.resamplings, dict):
+            avg = self.resamplings.get(k, 1)
+        else:
+            avg = self.resamplings
+
+        gradient = np.zeros(x.size)
+
+        # accumulate the number of samples
+        deltas = [bernoulli_perturbation(x.size, self.perturbation_dims) for _ in range(avg)]
```
Comment (nonhermitian) [inline qiskit/algorithms/optimizers/spsa.py:242]: this can likely be done using arrays.

File: qiskit/algorithms/optimizers/spsa.py [hunk 4]
```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
...
+            x = x_next
+
+            # update the list of the last ``last_avg`` parameters
+            if self.last_avg > 1:
+                last_steps.append(x_next)
+                if len(last_steps) > self.last_avg:
+                    last_steps.popleft()
+
+            # check termination
+            if np.linalg.norm(update) < self.tolerance:
```
Comment (nonhermitian) [inline qiskit/algorithms/optimizers/spsa.py:336]: why break if tolerance is true for a single iter given that everything is stochastic?  The next iter could be outside this tolerance.  Does it not make sense to track this as an ensemble average?

File: qiskit/algorithms/optimizers/spsa.py [hunk 5]
```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
...
+
+            # trust region
+            if self.trust_region:
+                norm = np.linalg.norm(update)
+                if norm > 1:  # stop from dividing by 0
+                    update = update / norm
+
+            # compute next parameter value
+            update = update * next(eta)
+            x_next = x - update
```
Comment (nonhermitian) [inline qiskit/algorithms/optimizers/spsa.py:296]: Do we not want to bound the variables in the range [0, 2pi] given that they are phase angles?

File: qiskit/algorithms/optimizers/spsa.py [hunk 6]
```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
...
+        # keep track of the last few steps to return their average
+        last_steps = deque([x])
+
+        for k in range(1, self.maxiter + 1):
+            iteration_start = time()
+            # compute update
+            update = self._compute_update(loss, x, k, next(eps))
+
+            # trust region
+            if self.trust_region:
```
Comment (nonhermitian) [inline qiskit/algorithms/optimizers/spsa.py:289]: why do we need this?  Doesn't the `c` parameter in the alg already control the maximum step that can be taken?

File: qiskit/algorithms/optimizers/spsa.py [hunk 7]
```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
...
+            if self.blocking:
+                fx_next = loss(x_next)
+
+                self._nfev += 1
+                if fx + self.allowed_increase <= fx_next:  # accept only if loss improved
+                    if self.callback is not None:
+                        self.callback(self._nfev,  # number of function evals
+                                      x_next,  # next parameters
+                                      fx_next,  # loss at next parameters
+                                      np.linalg.norm(update),  # size of the update step
```
Comment (nonhermitian) [inline qiskit/algorithms/optimizers/spsa.py:308]: The norm here depends on the dim of the vector.  Since the steps all the same, can one not just take the inf norm as the norm used?

**Review by nonhermitian — COMMENTED** [review]
Some minor comments.

File: qiskit/algorithms/optimizers/spsa.py (hunk 7 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/spsa.py:320]: Yes you could for standard SPSA. However that's not true anymore for 2-SPSA and QN-SPSA (see links in the description) since the gradient is multiplied with the inverse Hessian. Since we want to add these algos we already used the L2 norm here.

File: qiskit/algorithms/optimizers/spsa.py (hunk 4 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/spsa.py:348]: good point... maybe a tolerance is not a great idea for SPSA. I'll remove it until we come up with something better 👍🏻

File: qiskit/algorithms/optimizers/spsa.py (hunk 2 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/spsa.py:375]: it should, thanks!

File: qiskit/algorithms/optimizers/spsa.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/spsa.py]: because I've overwritten the whole file with the new implementation 😉 I'll revert to the old copyright years

File: qiskit/algorithms/optimizers/spsa.py (hunk 5 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/spsa.py:308]: I don't think we want to do this generally, since 
1. the angles can have coefficients 
2. it would imply you can only ever use SPSA for circuits (maybe you have a more generic loss function)
3. even if you only have circuits with coeffs equals to 1, it might be cheaper to "jump over" the 2pi boundary than "walk back", if that makes sense 

😄

File: qiskit/algorithms/optimizers/spsa.py (hunk 6 above)
Comment (nonhermitian) [reply qiskit/algorithms/optimizers/spsa.py:300]: one is rejecting steps that go to large, do we not want big downward changes to be accepted, i.e. don't we want to follow steep gradients downward?

File: qiskit/algorithms/optimizers/spsa.py (hunk 5 above)
Comment (nonhermitian) [reply qiskit/algorithms/optimizers/spsa.py:307]: Not sure what a angle coefficient is.  I was thinking of a limited usage model where all parameters correspond to 1Q gate angles, and thus are cyclic mod 2pi.  But perhaps there is another use case for quantum that I am missing?  I am thinking that the routine here is not for general problems, if one desires a generic solver then why bury it here and not expose it as a stand-alone package or try adding it to SciPy as a general purpose optimizer?

File: qiskit/algorithms/optimizers/spsa.py (hunk 7 above)
Comment (nonhermitian) [reply qiskit/algorithms/optimizers/spsa.py:319]: Ok.  I have no experience with the second-order method.

File: qiskit/algorithms/optimizers/spsa.py (hunk 5 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/spsa.py:307]: I mean that e.g this circuit is not cyclic mod 2 in `x` 🙂 
```
     ┌──────────┐
q_0: ┤ RY(x/10) ├
     └──────────┘
```

File: qiskit/algorithms/optimizers/spsa.py (hunk 6 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/spsa.py:300]: That can be achieved using the blocking condition (`blocking=True`), this attribute is really just to confine the size of the update step to improve stability in possibly too noisy functions.

**Review by woodsp-ibm — APPROVED** [review]

**Comment by gcattan** [comment]
Hi, I am migrating old code from qiskit-aqua.
With this refactoring, I am not sure how I should adapt the code below:
`SPSA(maxiter=1000, c0=0.1, c1 = 0.01)`
Could you help me?
Thanks!
